File: leetcode/PathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def checkSum(self,
                curr: TreeNode,
                accumulator : int,
                targetSum : int):

        if (curr == None):
            return False

        accumulator += curr.val

        if (accumulator == targetSum and curr.left == None and curr.right == None):
            return True

        return self.checkSum(curr.left,
                             accumulator,
                             targetSum) or self.checkSum(curr.right,
                                                         accumulator,
                                                         targetSum)

    def hasPathSum(self, root: TreeNode, targetSum: int) -> bool:

        if (root == None):
            return False

        if (root.val == targetSum and root.left == None and root.right == None):
            return True

        return self.checkSum(root.left,
                             root.val,
                             targetSum) or self.checkSum(root.right,
                                                         root.val,
                                                         targetSum)

File: leetcode/PathSum.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
    def checkSum(curr : TreeNode,
                 accumulator: Int,
                 targetSum : Int): Boolean = {


        if (curr == null) {
            return false;
        }

        // there's no += in scala

        if ((accumulator+curr.value) == targetSum && curr.left == null && curr.right == null) {
            return true;
        }

        return checkSum(curr.left, (accumulator+curr.value), targetSum) || checkSum(curr.right, (accumulator+curr.value), targetSum);


    }

    def hasPathSum(root: TreeNode, targetSum: Int): Boolean = {

        if (root == null){
            return false;
        }

        if (root.value == targetSum && root.left == null && root.right == null) {
            return true;
        }
        var accumulator =  root.value

        return checkSum(root.left, accumulator, targetSum) || checkSum(root.right, accumulator, targetSum);

    }
}
